src/eval/operation.rs

```diff
@@ -1851,15 +1964,53 @@ fn process_binary_operation(
 
                             let mut ts: Vec<RichTerm> = Vec::with_capacity(ts1.len() + ts2.len());
 
-                            ts.extend(ts1.into_iter());
-                            ts.extend(ts2.into_iter());
-
                             let mut env = env1.clone();
                             // TODO: Is there a cheaper way to "merge" two environements?
                             env.extend(env2.iter_elems().map(|(k, v)| (k.clone(), v.clone())));
 
+                            // We have two sets of contracts from the LHS and RHS arrays.
+                            // - Common contracts between the two sides can be put into
+                            // `pending_contracts` of the resulting concatenation as they're
+                            // shared by all elements: we don't have to apply them just yet.
+                            // - Contracts thats are specific to the LHS or the RHS have to
+                            // applied because we don't have a way of tracking which elements
+                            // should take which contracts.
+
+                            let ctrs_left = attrs1
+                                .pending_contracts
+                                .iter()
+                                .filter(|ctr| !attrs2.pending_contracts.contains(ctr))
+                                .cloned();
+
+                            let ctrs_right = attrs2
+                                .pending_contracts
+                                .iter()
+                                .filter(|ctr| !attrs1.pending_contracts.contains(ctr))
+                                .cloned();
+
+                            ts.extend(ts1.into_iter().map(|t|
+                                apply_contracts(t, ctrs_left.clone(), pos1)
+                                .closurize(&mut env, env1.clone())
+                            ));
+
+                            ts.extend(ts2.into_iter().map(|t|
+                                apply_contracts(t, ctrs_right.clone(), pos2)
+                                .closurize(&mut env, env2.clone())
+                            ));
+
+                            let ctrs_common = attrs1
```

Comment:
I think you can use `std::iter::partition` above to compute `ctrs_left` and `ctrs_common` at once above, instead of doing another pass here. Maybe that will saved one `cloned()` as well